### ISSUE TITLE
containerd: container not found when exec into container, but container is running fine (healthcheck is also failling) - Centos 7.4, docker 17.09.0-ce with overlay2 driver

### ISSUE THREAD

**Issue by jonesbusy** [moby/moby#35091]
**Description**

After some time a container becomes unhealthy (because of "containerd: container not found"). The container is running and accessible but it's not possible to "exec" into it.  The container is also marked "unhealthy" because the healhcheck cannot be executed inside the container (same message).

**Steps to reproduce the issue:**
1. Run a container (success)
2. Stop and remove container (success)
3. Build new image (success)
4. Start the image (success)
5. The container is accessible but is unhealty and it's not possible to "exec" into it

**Describe the results you received:**

```
# docker exec -it 22c0e5b6c034 /bin/bash
# containerd: container not found
```

```
22c0e5b6c034        9d09aee6d40c                                                   "/wait-for-it.sh -..."   5 hours ago         Up 5 hours (unhealthy)   9392/tcp                                              *****
```

Docker stats also return invalid data

```
docker stats --no-stream 22c0e5b6c034
```

```
CONTAINER           CPU %               MEM USAGE / LIMIT   MEM %               NET I/O             BLOCK I/O           PIDS
22c0e5b6c034        0.00%               0B / 0B             0.00%               0B / 0B             0B / 0B             0
```

```
docker restart 22c0e5b6c034
--> Wait indefinitely, the container stay in the same state...
```

```
docker stop 22c0e5b6c034
--> The command return, but the container stay in the same state
```

```
docker kill 22c0e5b6c034
--> The command return, but the container stay in the same state
```

```
docker network disconnect <network name> 22c0e5b6c034
The container is successfully disconnected from network. But still not possible to stop or kill container
```

```
docker rm -f 22c0e5b6c034
The container is finally removed (Not sure about the behavior of forcing the removal...)
```

**Describe the results you expected:**

Access to container and health check is running. The container can be restarted or stopped without restarting the docker daemon.

**Additional information you deem important (e.g. issue happens only occasionally):**

Issue happens only occasionally on different container

**Output of `docker version`:**

```
Client:
 Version:      17.09.0-ce
 API version:  1.32
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:41:23 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.0-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.3
 Git commit:   afdb6d4
 Built:        Tue Sep 26 22:42:49 2017
 OS/Arch:      linux/amd64
 Experimental: false

```

**Output of `docker info`:**

```
Containers: 92
 Running: 91
 Paused: 0
 Stopped: 1
Images: 642
Server Version: 17.09.0-ce
Storage Driver: overlay2
 Backing Filesystem: xfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 3.10.0-693.2.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 29.8GiB
Name: ******************
ID: VPH3:AQDV:6LGC:PPPT:72KB:TFRX:HS2L:G6EQ:6ZWV:W2QD:WMPA:3YIV
Docker Root Dir: /data/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Centos 7.4 running on AWS

**Comment by jonesbusy** [comment]
Any update on that issue ?

Still happening occassionaly..

```
            "Health": {
                "Status": "unhealthy",
                "FailingStreak": 566,
                "Log": [
                    {
                        "Start": "2017-10-13T06:04:21.592299257Z",
                        "End": "2017-10-13T06:04:21.614333267Z",
                        "ExitCode": -1,
                        "Output": "containerd: container not found"
                    },
                    {
                        "Start": "2017-10-13T06:04:51.622229814Z",
                        "End": "2017-10-13T06:04:51.630053038Z",
                        "ExitCode": -1,
                        "Output": "containerd: container not found"
                    },
                    {
                        "Start": "2017-10-13T06:05:21.635425644Z",
                        "End": "2017-10-13T06:05:27.099476322Z",
                        "ExitCode": -1,
                        "Output": "containerd: container not found"
                    },
                    {
                        "Start": "2017-10-13T06:05:57.122231607Z",
                        "End": "2017-10-13T06:05:59.369197033Z",
                        "ExitCode": -1,
                        "Output": "containerd: container not found"
                    },
                    {
                        "Start": "2017-10-13T06:06:29.378303279Z",
                        "End": "2017-10-13T06:06:29.393320519Z",
                        "ExitCode": -1,
                        "Output": "containerd: container not found"
                    }
```

**Comment by rodrigo-brito** [comment]
@jonesbusy same problem here. Restarting the docker solves the problem?

**Comment by ctritten** [comment]
Same problem here too.

Some of the containers are correctly running but impossible to interract with them (stop, exec...).

Docker stats returns no metrics for those containers.

The healthchecks are not working and logs the message:

msg="Health check for container 494f2d097482fe72c232bdf40e589170885a7109386f8b1d7cfbf8737dff263d error: containerd: container not found".

**Comment by rodrigo-brito** [comment]
I restart the docker and solved my problem. In the system log, I saw a possible memory dump.

**Comment by jonesbusy** [comment]
@rodrigo-brito Yes restarting the docker daemon fix the issue

**Comment by eugene-dounar** [comment]
Hi all! We also faced this issue on 17.09

In our case it's caused by containerd being killed by dockerd after failing healthchecks. Containerd leaves the container's state dir in inconsistent state. When it's being restarted it removes the dir assuming the container is dead
```
time="2018-02-21T16:00:45.737542634Z" level=warning msg="containerd: failed to load container,removing state directory." error="open /var/run/docker/libcontainerd/containerd/c4a964540cb5c0264b8a241caff725787be3b6917b9031549d5ebc795100fd72/d6fd3b57aee840873bd4321543f4f8f73b8d5e3ff46601c8915549adfa934540/process.json: no such file or directory" id=c4a964540cb5c0264b8a241caff725787be3b6917b9031549d5ebc795100fd72
```

Again, in our case containerd is killed after it created the "process dir" for `docker exec`, but before it writes `process.json` file. 

Symptoms:
- containerd is having health issues and being restarted by dockerd
- running multiple `docker exec` or docker healthchecks

On some machines we've seen this bug disappear after upgrading to 17.12, which uses new containerd version.

**Comment by ryanlyy** [comment]
I met this exactly same issue. why do we meet this issue? is there any bug? although restart dockerd can solve it but it is not expected solution. in living system, it is risk we can't always expect to restart dockerd.

**Comment by romanlevin** [comment]
Ran into this issue as well some excerpts from the daemon log:

```
Jun 21 20:15:40 egnyte_turbo dockerd[640]: time="2018-06-21T20:15:40.184549486Z" level=warning msg="containerd: failed to load container,removing state directory." error="open /var/run/docker/libcontainerd/containerd/99a575e79f173c8b90e5e
Jun 21 20:15:56 egnyte_turbo dockerd[640]: time="2018-06-21T20:15:56.880517106Z" level=warning msg="Health check for container 99a575e79f173c8b90e5ef2bd6fb605da7b1d2055cd7a21615c2d596f611b31d error: containerd: container not found"
Jun 21 20:18:35 egnyte_turbo dockerd[640]: time="2018-06-21T20:18:35.334535115Z" level=warning msg="container kill failed because of 'container not found' or 'no such process': Cannot kill container 99a575e79f173c8b90e5ef2bd6fb605da7b1d20
```

**Comment by EslamElHusseiny** [comment]
Same issue here
```
# docker version
Client:
 Version:      17.09.1-ce
 API version:  1.32
 Go version:   go1.8.4
 Git commit:   3dfb8343b139d6342acfd9975d7f1068b5b1c3d3
 Built:        Mon Jan  8 22:44:09 2018
 OS/Arch:      linux/amd64

Server:
 Version:      17.09.1-ce
 API version:  1.32 (minimum version 1.12)
 Go version:   go1.8.4
 Git commit:   402dd4a/17.09.1-ce
 Built:        Mon Jan  8 22:44:42 2018
 OS/Arch:      linux/amd64
 Experimental: false

cat /etc/os-release
NAME="Amazon Linux AMI"
VERSION="2017.09"
ID="amzn"
ID_LIKE="rhel fedora"
VERSION_ID="2017.09"
PRETTY_NAME="Amazon Linux AMI 2017.09"
ANSI_COLOR="0;33"
CPE_NAME="cpe:/o:amazon:linux:2017.09:ga"
HOME_URL="http://aws.amazon.com/amazon-linux-ami/"
```

**Comment by cpuguy83** [comment]
Closing since this is fixed since 17.12.
18.03 is the latest supported release.
